rmw_fastrtps_shared_cpp/src/rmw_get_topic_info.cpp

```diff
@@ -0,0 +1,311 @@
+  auto & topic_cache =
+    is_publisher ? slave_target->writer_topic_cache : slave_target->reader_topic_cache;
+  {
+    std::lock_guard<std::mutex> guard(topic_cache.getMutex());
+    const auto & topic_name_to_data = topic_cache().getTopicNameToTopicData();
+    std::vector<rmw_topic_info_t> topic_info_vector;
+    for (const auto & topic_name : topic_fqdns) {
+      const auto it = topic_name_to_data.find(topic_name);
+      if (it != topic_name_to_data.end()) {
+        for (const auto & data : it->second) {
+          rmw_topic_info_t topic_info = rmw_get_zero_initialized_topic_info();
+          rmw_ret_t ret = _set_rmw_topic_info(
+            &topic_info,
+            participant_guid,
+            node_name,
+            node_namespace,
+            data,
+            no_mangle,
+            slave_target,
+            allocator);
+          if (ret != RMW_RET_OK) {
+            // Free topic_info
+            _handle_topic_info_fini(&topic_info, allocator);
+            // Free all the space allocated to the previous topic_infos
+            for (auto & tinfo : topic_info_vector) {
+              _handle_topic_info_fini(&tinfo, allocator);
+            }
+            return ret;
+          }
+          // add rmw_topic_info_t to a vector
+          topic_info_vector.push_back(topic_info);
+        }
+      }
+    }
+
+    // add all the elements from the vector to rmw_topic_info_array_t
+    auto count = topic_info_vector.size();
+    ret = rmw_topic_info_array_init_with_size(participants_info, count, allocator);
+    if (ret != RMW_RET_OK) {
+      rmw_error_string_t error_message = rmw_get_error_string();
+      rmw_reset_error();
+      RMW_SET_ERROR_MSG_WITH_FORMAT_STRING(
+        "rmw_topic_info_array_init_with_size failed to allocate memory: %s", error_message.str);
+      // Free all the space allocated to the previous topic_infos
+      for (auto & tinfo : topic_info_vector) {
+        _handle_topic_info_fini(&tinfo, allocator);
+      }
+      return ret;
+    }
+    for (auto i = 0u; i < count; i++) {
+      participants_info->info_array[i] = topic_info_vector.at(i);
+    }
+    participants_info->count = count;
+  }
+  return RMW_RET_OK;
+}
+
+rmw_ret_t
+__rmw_get_publishers_info_by_topic(
+  const char * identifier,
+  const rmw_node_t * node,
+  rcutils_allocator_t * allocator,
+  const char * topic_name,
+  bool no_mangle,
+  rmw_topic_info_array_t * publishers_info)
+{
+  return _get_info_by_topic(
+    identifier,
+    node,
+    allocator,
+    topic_name,
+    no_mangle,
+    true, /*is_publisher*/
+    publishers_info);
+}
+
+rmw_ret_t
+__rmw_get_subscriptions_info_by_topic(
+  const char * identifier,
+  const rmw_node_t * node,
+  rcutils_allocator_t * allocator,
+  const char * topic_name,
+  bool no_mangle,
+  rmw_topic_info_array_t * subscriptions_info)
+{
+  return _get_info_by_topic(
+    identifier,
+    node,
+    allocator,
+    topic_name,
+    no_mangle,
+    false, /*is_publisher*/
```

Comment:
This style of reuse lends something to my question in the rmw pr about whether or not there should be two rmw functions or not...

Reply:
If we would have just one function for both publishers and subscriptions, we would still need a function like the above one, call it twice and aggregate results. That's because `publishers` and `subscriptions` are stored in different topic cache objects:
https://github.com/ros2/rmw_fastrtps/blob/c231f14f92759a898c05af27b2a56a97c44038ae/rmw_fastrtps_shared_cpp/src/rmw_get_topic_info.cpp#L218

Reply:
Well, that's kinds of an implementation detail, it could also be a single function like this:

```
...
rmw_get_topic_info(...)
{
    ... result;

    for (... entry : slave_target->writer_topic_cache) {
      // add entry to result, could use function for adding single item to result
    }

    for (... entry : slave_target->reader_topic_cache) {
      // add entry to result
    }

    return result;
}
```

Reply:
We've decided on sticking to two `rmw` APIs.